### DISCUSSION TITLE
How to start Quasar app manually?

### DISCUSSION THREAD

**Discussion by fairking** [quasarframework/quasar#13725] [CLI - PWA mode]
I am looking for a solution to start Quasar Manually. The reason is some of the api function needs to be initialized first before quasar boot. Does anybody know if it's possible??

something like `quasar.config.js` (note, this is a dummy setting):
``` js
module.exports = configure(function () {
    return {
        autostart: false;
    };
}
```

and then `my.api.initializer.js` (note, this is a dummy function):
``` js
function on_api_initialized() {
    qApp.boot();
}
```

As a workaround I can try to load the js into DOM, but in my opinion this would be very dirty approach:
``` js
function on_api_initialized() {
    var jq = document.createElement('script');
    jq.src = "/index.xxxxxx.js";
    document.getElementsByTagName('head')[0].appendChild(jq);
}
```

By the way, I am not using Capasitor or Electron. Just pure PWA which goes into MAUI web container.

**Comment by TobyMosque** [comment]
u can create a boot who verifies if the API is up and running.
```js
import { axios } from 'axios'
import { boot } from 'quasar/wrapper'

export default boot(async () => {
  const api = axios.create({ /*...*/ })
  let backoff = 500
  do {
    try {
      await api.get('wealth')
      break;
    } catch {
      await new Promise(resolve => setTimeout(resolve, backoff))
      backoff *= 1.5
    }
  } while (true)
})
```

on better, learn docker and make your app depends on your api.

**Reply by fairking** [reply]
> on better, learn docker and make your app depends on your api.

This is a desktop/mobile offline app. The API is on the device, not in the web. (Backend API - C# .NET).

**Comment by TobyMosque** [comment]
Maybe I'm wrong, but i think u trying to build a app with blazor, unfortunally i never tried that myself, and personally i don't recommend that approach.

In your place, i would use the Dotnet CLI only to develop/build the backend (gPRC, REST or GraphQL).
And use the Quasar CLI to developer the frontend (Electron to desktop and Capacitor to Mobile).
If u really wanna to try something new, maybe u would look into Tauri (https://tauri.app).

But if u really wanna to stick with Blazor, maybe u would look for help in the Blazor community.

**Reply by fairking** [reply]
Thanks, I very appreciate your opinion. 

Do you know by any chance what is the tauri's backend language? Maybe I can look at that, as Blazor has some issues and looks like they are not going to solve them. So for the hybrid apps Blazor could not be an option.

**Reply by fairking** [reply]
BTW the .NET CLI you mentioned is working only on desktop, but not on mobile. As far as I know.

**Reply by fairking** [reply]
> Do you know by any chance what is the tauri's backend language?

Ok I have found it is Rust actually. Very nice backend language. I should probably try it out definitely. https://github.com/tauri-apps/smoke-tests

**Reply by TobyMosque** [reply]
Tauri is just a bridge made in Rust who connects the system webview and the operation system. that works very like ElectronJS.
Tauri is frontend agnostic, so u can use with any framework (vue, react, angular, quasar, etc)
Tauri is backend agnostic, so u can connect to any backend (ASP.NET, NestJS, etc) using any protocol (gRPC, REST, GraphQL).

If u really wanna a more monolitic approach, one single app who encapsulastes the frontend (C#) and the backend (C#), maybe u would look into the MAUI:
https://github.com/dotnet/maui
https://devblogs.microsoft.com/dotnet/introducing-dotnet-maui-one-codebase-many-platforms/

**Reply by fairking** [reply]
Thanks Toby, but I think c# (Razor) is bad for frontend development. I will look at the Tauri.

**Reply by TobyMosque** [reply]
MAUI has a different approach, that is based in the XAML (what i dislike), so that is more of a flutter (native cross platform development) alternative then a quasar one (hybrid development with Js/html/css)

**Reply by fairking** [reply]
@TobyMosque I can explain you a little bit more. MAUI is a new cross platform framework which is more than XAML. You can literally write Hybrid applications using [Blazor](https://github.com/dotnet/blazor-samples/blob/main/6.0/BlazorWebAssemblySignalRApp/Client/Pages/Counter.razor). Moreover their BlazorWebView is very nice as it has a backend(c#)<=>frontend(js) communication. So you can literally call a backend from js and another way around. 

I managed to stick the Quasar app into the view and speak to backend in the same way as I was speaking to web api (think axios). So everything is working fine so far, the only that startup issue I have, which I can workaround, but I am looking for some proper more professional solution.
I can run the app as web or as desktop(windows/linux/macos) or mobile (android/ios). Same UI. Same backend. Tested on physical devices. Everything is working fine.
